### PR TITLE
Update Algebra Homomorphisms

### PR DESCRIPTION
**Pull request by raulepure**
Changed AlgHom type by replacing the lex ordering in the computations by a dp*dp resp. dp*lex ordering.
Added groebner_assure with product orderings.
Replaced mapping of elements via Singular.jl by using MapFromFunction and evaluate.
Added check of homoegeneity in case codomain is homogeneous.

### CODE REVIEW

**Comment by tthsqe12** [comment]
Working on it. related https://github.com/oscar-system/Oscar.jl/pull/354

File: src/Rings/affine-algebra-homs.jl [hunk 1]
```diff
@@ -84,20 +84,69 @@ mutable struct AlgHom{T} <: AbstractAlgebra.Map{Ring, Ring,
...
+   W <: Union{MPolyRing{T}, MPolyQuo{S}},
+   X <: Union{S, MPolyQuoElem{S}}}
 
-   function AlgHom{T}(R::U, S::W, V::Vector{X}) where {T, U, W, X}
-      Rx = singular_ring(R)
-      Sx = singular_ring(S)
-      z = new(R, S, V, Singular.AlgebraHomomorphism(Rx, Sx, Sx.(V)))
+      if typeof(C) <: MPolyRing_dec
+         bit = ishomogenous.(V)
+         length(findall(x->x==1, bit)) == length(V) || error("Array needs to contain homogeneous elements")
```
Comment (tthsqe12) [inline src/Rings/affine-algebra-homs.jl]: Isn't this the same thing as `all(ishomogenous, V) || error(...)`

File: src/Rings/affine-algebra-homs.jl [hunk 2]
```diff
@@ -84,20 +84,69 @@ mutable struct AlgHom{T} <: AbstractAlgebra.Map{Ring, Ring,
...
+                       # the ideal generated by I and f_i-y_i in T with its respective groebner data
+                       # and the data D of the division with remainder of the x_i w.r.t G.
+
+   function AlgHom{T}(D::MPolyRing{T}, C::W, V::Vector{X}) where {T, S <: MPolyElem{T},
+   W <: Union{MPolyRing{T}, MPolyQuo{S}},
+   X <: Union{S, MPolyQuoElem{S}}}
+      
+      if typeof(C) <: MPolyRing_dec
+         bit = ishomogenous.(V)
+         length(findall(x->x==1, bit)) == length(V) || error("Array needs to contain homogeneous elements")
```
Comment (tthsqe12) [inline src/Rings/affine-algebra-homs.jl]: ditto

**Review by tthsqe12 — COMMENTED** [review]
There is a lot here, and I can just make trivial comments at this time. I will try to go more indepth tomorrow.

**Comment by tthsqe12** [comment]
Do the barriers to makeing AlgHom parametrized by the domain and codomain element types still exist?

**Comment by raulepure** [comment]
There are still some parts which are problematic like using the base_ring function on an quotient ring and on a graded ring. In one case you get the underlying polynomial ring in the other case you get the base field. 
And I also had some trouble with evaluate in this context,.
This yields to the separation of these types in the code. 

If it is what you mean.

**Comment by thofma** [comment]
(We can now use `coefficient_ring` to get the underling coefficients for graded rings. Not sure if this helps though.)

**Comment by tthsqe12** [comment]
Currently the `T` in `AlgHom{T}` is the type of the coefficients in the domain and codomain. This makes writing code against `AlgHom` more difficult, and, more importantly, omits maps from `Z[x]` to `F_5[z]`. As long as all of the cases that are supposed to work are tested in the test code, I would say this can be merged, and we can extent the `AlgHom` in a later pr using the new `coefficient_ring` standard. Also, unless I am missing something,
```julia
bit = ishomogenous.(V)
length(findall(x->x==1, bit)) == length(V)
```
is a much more confusing way of writing
```julia
all(ishomogenous, V)
```

File: src/Rings/affine-algebra-homs.jl (hunk 1 above)
Comment (tthsqe12) [inline src/Rings/affine-algebra-homs.jl:144]:
```suggestion
all(ishomogenous, V) || error(...)
```

Comment (tthsqe12) [inline src/Rings/affine-algebra-homs.jl]:
```suggestion
         all(ishomogenous, V) || error("Array needs to contain homogeneous elements")
```

File: src/Rings/affine-algebra-homs.jl (hunk 2 above)
Comment (tthsqe12) [inline src/Rings/affine-algebra-homs.jl]:
```suggestion
         all(ishomogenous, V) || error("Array needs to contain homogeneous elements")
```

**Comment by jieaosong** [comment]
Just want to comment on the spelling of `ishomogeneous`...

**Comment by thofma** [comment]
Ah yes, thanks. Obviously this code is broken since `ishomogenous` does not exist. Can you fix the spelling and add a test @tthsqe12?

**Comment by raulepure** [comment]
Thank you for changing the code. @tthsqe12 @thofma 
I fixed another "bit ..." into "all(ishomogeneous, V)" which was overlooked and I added a small test. 
The PR is coming in some minutes, I'm just running the test

**Pull request merged**